### DISCUSSION TITLE
Folding / unfolding nodes changes focus to selected node

### DISCUSSION THREAD

**Discussion by mhechthz** [freeplane/freeplane#1190] [Help]
In the new Freeplane 1.11.2 the focus/selection on nodes is not longer by hovering the mouse over the node but by clicking on it. On navigating in the map everytime I fold or unfold a node that is not selected, the focus jumps back to the selected node. This is very annoying, since I always need to click on the node befor folding/unfolding.

How can the focus change be deactivated?
(and how can selection by hovering be activated again)

**Comment by macmarrum** [comment]
Just to cover all bases, what is set in `Preferences…->Behaviour->Selection method->Selection method`?

**Comment by mhechthz** [comment]
Ok, I checked all settings but missed this one. So selection on hover works. Thank you!

Nevertheless, in "select by click" mode  the problem of always focus jumping to selected node on folding/unfolding other nodes remains. Is this a bug or a feature. And if a feature: why, and how to deactivate?

EDIT: The annoying focus jump is also active with "delayed select" mode, but since in most cases the fold/unfold click takes longer than the delay, it is no problem (as long as the delay is not set to higher values, e.g. 1s).

**Comment by macmarrum** [comment]
I cannot see this behavior on my Freeplane 1.11.2.

What operating system and Java version are you using? _You can find Java version in `Help->About`_

1. Does it happen for each mindmap, i.e. if you create a new one, can you observe the same behavior?
2. Would it be OK for you to try with a fresh profile? Here are the steps:
    - `Tools->Open user directory`
    - go one level up, so that you can see the folder `1.11.x`
    - close Freeplane
    - rename the folder `1.11.x` to e.g. `1.11.x-original`
    - start Freeplane and try it now

To revert to your original profile, close Freeplane, delete the newly created `1.11.x` and rename `1.11.x-original` to `1.11.x`.

**Comment by mhechthz** [comment]
Here is the "About" diolog (it's in German):
![image](https://github.com/freeplane/docs/assets/37075144/f64cb4e5-4200-4f09-bf0e-2690ff39e814)

And here are my settings (also in German):
![image](https://github.com/freeplane/docs/assets/37075144/e48a90c6-c391-4e4f-9ce2-467af77acc6f)
It says: delayed, 50 ms, do not center selected node, do not scroll slowly to selected node

I'm on Windows 11.

Yes, also on new maps it has this behavior.
![image](https://github.com/freeplane/docs/assets/37075144/24513c3f-40ac-49eb-97a7-3b083da7fd5c)
E.g. this map. If I move the `baa `node out of the window and fold/unfold the `aa `node then the `baa `node is always moved to the window again.

a) Start view
![image](https://github.com/freeplane/docs/assets/37075144/98fa9adb-aebd-4b18-bd3a-c372e8d7eb8a)

b) Just unfolding `aa` --> new view
![image](https://github.com/freeplane/docs/assets/37075144/dda9dd29-dffb-42e7-a914-670eeea4f93d)

For large maps this is barely acceptable.

EDIT: On zooming by CTRL + mouse wheel to a not selected node the focus jumps also to the selected node. This is absolutely unwanted. Even on deactivating "zoom to selected node" the zoom doen's follow the mouse pointer.

**Comment by macmarrum** [comment]
Thanks. Now I can understand better what you mean. I can see the same behavior in my Freeplane 1.11.2.

@dpolivaev Could you please have a look?

**Comment by dpolivaev** [comment]
Dear @mhechthz

So many people are trying to help you, to understand your case and to find a solution just for the sake of helping. I have been working on the project for free since more than 20 years.

Do you really have no words for us other than "barely acceptable"?

**Comment by mhechthz** [comment]
Sorry! I really like Freeplane. You have done amazing work on this! Thank you in the name of community. :-)

Sorry for my coarse and unpolished english.

**Comment by mhechthz** [comment]
I downloaded a portable version of Freeplane 1.10.6. It seems, that the behavior here is different to 1.11.2 and as expected.

**Comment by mhechthz** [comment]
The "not zoom to mouse pointer" issue i mentioned above could be corrected by the right settings in the configuration menu.

**Comment by mhechthz** [comment]
After fiddling around with some settings and mindmaps I get the impression, that the focus jump to selected node sometimes happens (most of the time) and sometimes doesn't happen.

**Comment by dpolivaev** [comment]
> Sorry! I really like Freeplane. You have done amazing work on this! Thank you in the name of community. :-)

I am glad that you like it. You are welcome to the community. However neither you nor I can speak in the name of community because we have not been elected to represent the community. Here everyone makes points only in his own name.

**Comment by dpolivaev** [comment]
> I downloaded a portable version of Freeplane 1.10.6. It seems, that the behavior here is different to 1.11.2 and as expected.

> After fiddling around with some settings and mindmaps I get the impression, that the focus jump to selected node sometimes happens (most of the time) and sometimes doesn't happen.

Who knows what is expected? Different people can have different expectations.
What do you mean saying "the focus jumps to the selected node"?

Having a focus means the same as being selected, namely that the node is the target of all user actions.
I expect that this node should always be inside the visible area and that all exceptions from this rule should be eliminated. Otherwise, there is a risk of making unintentional changes to the map without noticing it.

**Comment by dpolivaev** [comment]
By the way, if you have any questions about Freeplane I think you could use the discussion forum instead  of creating an issue.

**Comment by dpolivaev** [comment]
Anyway, I changed the implementation so that the selected node is not made visible after folding if it was not on screen before.

**Comment by Bricklam** [comment]
Thanks.
my problem with the new behaviour (keeping the selected node in the viewport, instead of the unfolded node (node that is being acted on): it doesn't fit my use case:
since I have many & big maps, I need to unfold many nodes as quickly as possible (one by one using the mouse),
to see the children and to recognize & view the map content (to get familiar with the map and the important  nodes) (specially when the ascendant/parent nodes title doesn't always indicates clearly what children content is
(so  you have to unfold it to see what type/content of children it has)

So, the point is: if I had to select everynode before unfolding it: it would take more time & slow my fast navigation and make it more cumbersome. I think the ability of browsing the map quickly by the mouse is a big advantage.

and I'm also used to this behaviour in all the older Freeplane version since 1.3.x


I understand your argument: that the new behaviour is better to avoid making unintentional changes to the map without noticing it.

So, I have two suggestion to solve this issue:

First:
to add extra option (that allow the users to choose the best behaviour for them, based on their use cases)

Second:
or: to disable this new behaviour when the (viewing mode) is enabled:
since I usually implement (Edit > Viewing) mode anway before browsing the map by the mouse (to hide oval handles and to avoid the accidental clicking on them).
 So, if this suggestion is implemented: no risk of unintentional changes to the map, because the map is already in the viewing mode, and even if the selected node is not in the viewport, no change can happen to the folded/unfolded node

Best regards

**Comment by Bricklam** [comment]
I want also to add another important point:
my main issue with Freeplane that it is not as mouse friendly (in terms of navigation), as (Freemind) for example.
what I mean: since I'm a heavy mouse user: I want to fold/unfold nodes as fast as possible and navigate the map quickly with the mouse,
but there are two points that annoy me when navigating maps with Freeplane by the mouse:

(1) I cannot just fold/unfold a node by one double click on a non-selected node  (like Freemind does), but I have to make two double clicks in Freeplane  (or double click + single click), Where the first one should select the node, and the next click to fold/unfold the node,
 (two double clicks just to fold/unfold a node, which is very annoying for a mouse user), (specially when the option (Edit on double click) is disabled from the Preferences)
  but in Freemind: only one double click will select AND fold/unfold the target node at the same time.

I have a suggestion here that can solve this problem: if the Preference option: (Behaviour > Edit on double click) is disabled, then make
the double click to directly select + fold/unfold the node at the same time (instead of just selecting it)


(2) the oval handle in Freeplane, which always overlaps with the folding/unfolding process done by the mouse,
which can cause unintentional/unnoticed changes to the map, making mouse navigation uneasy process.

and I see that Freemind has a better solution for this problem, which can be implemented in Freeplane:
 when the user hover over the target node by the mouse:
the popping up folding symbol will have a big size, to be able to click on it easily by the mouse (without confusing it with the oval handle),
but in (Freeplane), the folding symbol is not big enough (and there is no way to change its size from Preferences, like the fixed folding symbol), so, there is a higher chance to click on the oval handle by mistake, because both are very close and have the same size.

I thinkk if these two issues are fixed: Freeplane will be a mouse friendly and it will become much easier and faster to view and navigate a map in Freeplane.

My goal is to make folding/unfolding nodes by the mouse as fast and as quick as possible. to increase productivity
(since folding/unfolding process can be considered the most frequent task for a Freeplane user)

Thanks

**Reply by dpolivaev** [reply]
I find it very useful that you are describing the problem you have instead of trying to propose a solution. Generally one of the biggest challenges in Freeplane development ist that users generally prefer to propose solutions instead of describing the actual needs.

1. Try to deactivate option Preferences…->Behaviour->Edit on double click which is responsible for the delay in folding/unfolding a node by a single click and let me know if it helps and if it resolves your problem.
2.  From this message, I learned that the folding circle can not be used for folding in the case when the drag handle of the child node overlaps it. Now I am trying to find a solution for this issue too.

**Reply by dpolivaev** [reply]
Actually, I can even use double click for select and fold/unfold if it is not used for editing

**Reply by dpolivaev** [reply]
I fixed the issues reported in your recent message, please try preview version 1.11.5_04 from https://sourceforge.net/projects/freeplane/files/freeplane%20preview/ and let me know if helps.

**Comment by Illegit** [comment]
I personally find it better. Thanks.

**Reply by dpolivaev** [reply]
I like it too. Clear problem description helped me a lot to improve it.

**Comment by Bricklam** [comment]
Thanks very much for your amazing work & for caring about your users concerns @dpolivaev
I didn't expect that you are going to fix these issues very quickly. you are really an amazing developer.

I just tested (1.11.5_04) and 
I found that two big annoying issues have been solved:
- Prevent overlapping node drag handler from stealing the mouse pointer
- Fast fold/unfold on node click if editing is not possible or editing on double click is disabled

and now using the mouse is much much easier, and map navigation is a better
I'm so happy about it.


I just discovered a minor bug in (1.11.5_pre04) and I also have one suggestion, that I addressed earlier:

1- for the suggestion: related to keeping the selected node in the viewport while folding/unfolding other node:
 how about adding an option that Disable/Enable this behaviour (if the map is in viewing mode and no worry for the map to change unintentionallly),
 to allow the users to have control over the desired behaviour, for example an option like this would be a good idea:
-- Fodling/Unfolding (non-selected node) keeps selected node visible (in Viewing mode)

2- for the bug: I just found a minor bug related to fodling/unfolding node by the mouse in (1.11.5_pre04) and I reported it here:
https://github.com/freeplane/freeplane/issues/1203

Best Regards

**Comment by dpolivaev** [comment]
Please check preview 1.11.5_05. The behavior is controlled by option Preferences…->Behaviour->Freeplane keeps selected node visible which now also controls what happens if you zoom the map in or out.

**Comment by Bricklam** [comment]
Yes. I just checked it and it works perfectly
(super easy folding/unfolding process, and I can have control over keeping the selected node in the view or not, and the bug is fixed.

Thank you very much for your great work.
Best wishes
